### PR TITLE
Refactor Servo velocity bounds enforcement. Disable flaky unit tests.

### PR DESCRIPTION
**Pull request by vatanaksoytezer**
Port of https://github.com/ros-planning/moveit/pull/2471 and an attempt to fix flakiness (again). Overall this PR:

- Ports 2471.
- Fixes most flakiness of the tests.
- Adds low latency (named basic_servo_tests)

### CODE REVIEW

**Comment by vatanaksoytezer** [comment]
Seems folks at nav2 had a similar case of flaky tests (see https://github.com/ros-planning/navigation2/pull/584). They solved it with seperating the test with the others.

In a related note this PR should solve the test failures related to servo_server not being opened.

**Comment by vatanaksoytezer** [comment]
This also includes #352

**Comment by vatanaksoytezer** [comment]
@JafarAbdi @AndyZe @tylerjw can you guys review this? It seems to pass for some time with multi thread and sleeps added to tests.

File: moveit_ros/moveit_servo/test/basic_servo_tests.cpp
Comment (tylerjw) [inline moveit_ros/moveit_servo/test/basic_servo_tests.cpp:96]: De-referencing the message here will cause an unnecessary copy.

```suggestion
    twist_stamped_pub->publish(msg);
```

File: moveit_ros/moveit_servo/test/launch/test_servo_collision.test.py
Comment (tylerjw) [inline moveit_ros/moveit_servo/test/launch/test_servo_collision.test.py:21]: 5 seconds seems like a long time to sleep.  I'll need to test this to understand better what it is doing.

**Review by tylerjw — COMMENTED** [review]
I glanced through the changes and left a couple comments.  Thank-you for working on this.  I'll test this soon and give you a more detailed review.

File: moveit_ros/moveit_servo/test/launch/test_servo_collision.test.py
Comment (vatanaksoytezer) [reply moveit_ros/moveit_servo/test/launch/test_servo_collision.test.py:21]: I thought the same, but I cannot reproduce the flaky errors in my local. Tested out 2 and 5 seconds, was still failing at 2. Passed 6/6 with 5 seconds sleep. The problem seems to be a concurrency issue where the test result file would not be generated yet when gtest looks up for it.

File: moveit_ros/moveit_servo/test/basic_servo_tests.cpp
Comment (vatanaksoytezer) [reply moveit_ros/moveit_servo/test/basic_servo_tests.cpp:96]: This would not compile as I declared msg from `moveit::util::make_shared_from_pool<control_msgs::msg::JointJog>();` which is a sharedPtr. However I can declare it as `control_msgs::msg::JointJog msg;` and do the same thing.

Comment (vatanaksoytezer) [reply moveit_ros/moveit_servo/test/basic_servo_tests.cpp:96]: Sending the new commit right now with this change. Appreciate the review, thanks!

Comment (tylerjw) [reply moveit_ros/moveit_servo/test/basic_servo_tests.cpp]: That is because the approach of `moveit::util::make_shared_from_pool` won't work for ros2 because the publish interface only takes either the message type itself or a unique_ptr to the message type.  You can try changing `make_shared_from_pool` to create `unique_ptr`.  Here is the documentation on it: https://docs.ros.org/en/foxy/Tutorials/Intra-Process-Communication.html

I'm sorry I didn't think of this before.  It might also might be worth removing `moveit::util::make_shared_from_pool` from the ros2 version of this until we figure out and benchmark if it is worthwhile to generate unique_ptr with this and just use `std::make_unique` for now.

Comment (vatanaksoytezer) [reply moveit_ros/moveit_servo/test/basic_servo_tests.cpp]: Learned something new today, thanks for the clarrification. Should I open a issue / PR to remove / uncomment `moveit::util::make_shared_from_pool`? Or I can open one up for benchmarking it to unique_ptr?

Comment (AndyZe) [reply moveit_ros/moveit_servo/test/basic_servo_tests.cpp]: Seems like we would probably want to just delete it. I'd open an issue for that.

Benchmarking would be interesting if you have tons of free time    ;)

**Comment by AndyZe** [comment]
I think you should rename the PR because it's more than just porting that PR now. Maybe something like:

`Refactor Servo velocity bounds enforcement. Fix flaky unit tests.`

File: moveit_ros/moveit_servo/test/launch/servo_launch_test_common.py
Comment (AndyZe) [inline moveit_ros/moveit_servo/test/launch/servo_launch_test_common.py:171]: Good job figuring this out! I'm really impressed by your persistence with that.

File: moveit_ros/moveit_servo/test/launch/test_servo_collision.test.py
Comment (AndyZe) [inline moveit_ros/moveit_servo/test/launch/test_servo_collision.test.py:4]: nitpick:  please alphabetize. Put `import time` before `import unittest`

File: moveit_ros/moveit_servo/test/launch/test_servo_integration.test.py
Comment (AndyZe) [inline moveit_ros/moveit_servo/test/launch/test_servo_integration.test.py:4]: alphabetize

File: moveit_ros/moveit_servo/test/launch/test_servo_pose_tracking.test.py
Comment (AndyZe) [inline moveit_ros/moveit_servo/test/launch/test_servo_pose_tracking.test.py:6]: alphabetize

File: moveit_ros/moveit_servo/test/launch/test_servo_singularity.test.py
Comment (AndyZe) [inline moveit_ros/moveit_servo/test/launch/test_servo_singularity.test.py:4]: ...

File: moveit_ros/planning_interface/test/launch/move_group_ompl_constraints.test.py
Comment (AndyZe) [inline moveit_ros/planning_interface/test/launch/move_group_ompl_constraints.test.py:4]: ...

File: moveit_ros/moveit_servo/test/launch/test_basic_servo.test.py
Comment (AndyZe) [inline moveit_ros/moveit_servo/test/launch/test_basic_servo.test.py:17]: Nitpick: unless this is a standard name, I think the naming is a little odd. Maybe something like:

`GTestWaitForCompletion`?

Please add a comment about what it does, too.

Also, can you move it into `servo_launch_test_common.py`? Since this same class appears several places.

File: moveit_ros/moveit_servo/test/launch/test_basic_servo.test.py
Comment (AndyZe) [inline moveit_ros/moveit_servo/test/launch/test_basic_servo.test.py:24]: Same comments apply to this class.

**Review by AndyZe — COMMENTED** [review]
Generally looks great!

File: moveit_ros/moveit_servo/test/launch/servo_launch_test_common.py
Comment (vatanaksoytezer) [reply moveit_ros/moveit_servo/test/launch/servo_launch_test_common.py:171]: Thanks! I think this line helps but unfortunately without the sleep statements we still have some flakiness.

File: moveit_ros/moveit_servo/test/launch/test_basic_servo.test.py
Comment (vatanaksoytezer) [reply moveit_ros/moveit_servo/test/launch/test_basic_servo.test.py:17]: Sure, I've put a comment, changed the name and moved it to `servo_launch_test_common.py`

File: moveit_ros/moveit_servo/test/launch/test_servo_collision.test.py
Comment (vatanaksoytezer) [reply moveit_ros/moveit_servo/test/launch/test_servo_collision.test.py:4]: That's fair. Changed it.

File: moveit_ros/moveit_servo/include/moveit_servo/servo.h
Comment (vatanaksoytezer) [reply moveit_ros/moveit_servo/include/moveit_servo/servo.h]: 👍🏻

File: moveit_ros/moveit_servo/test/launch/test_basic_servo.test.py
Comment (vatanaksoytezer) [reply moveit_ros/moveit_servo/test/launch/test_basic_servo.test.py]: Moving these classes to servo_launch_test_common seemed to break. I reverted back into files.

**Comment by gavanderhoorn** [comment]
@vatanaksoytezer: a small comment on logistics: it would be OK to `git push` commits in batches, instead of individually.

Pushing individual commits leads to the CI infrastructure starting jobs every time you push, and it also leads to everyone who *Watches* this repository receiving emails for every individual push as well.

Over the past few days, I've received 48 notifications about this PR alone, which seems a bit excessive.

**Comment by vatanaksoytezer** [comment]
@gavanderhoorn thanks for the heads up and sorry that you've received so many notifications. Although I am aware that I should not push that much, for this PR I was actually pushing on purpose to trigger actions and see if the flakiness on tests continue. I cannot reproduce the errors in my local machine and wanted to try as much as I can to see if it breaks with those changes.

On your comment I could probably test this in my own Github repository before opening the PR (which I did for the first 15 or so commits), but I found out tests be still failing. Will try my best to not push that much again.

**Comment by gavanderhoorn** [comment]
>  I was actually pushing on purpose to trigger actions and see if the flakiness on tests continue.

of course if you're trying to debug the GHA cfg, that would be a different situation.

It was more of a general comment though. The same happened in a few other PRs.

> I could probably test this in my own Github repository before opening the PR (which I did for the first 15 or so commits)

that seems like a good approach indeed.

**Comment by vatanaksoytezer** [comment]
> It was more of a general comment though. The same happened in a few other PRs.
> 
> >

No worries, I will be more careful with my other PRs. Thanks for giving the heads up.

**Review by AndyZe — APPROVED** [review]

**Comment by vatanaksoytezer** [comment]
@AndyZe @tylerjw @JafarAbdi I am thinking about disabling the tests again since CI still seems to fail with the same errors after my additions.

**Comment by AndyZe** [comment]
> I am thinking about disabling the tests again since CI still seems to fail with the same errors after my additions.

Yep, I guess that is what we have to do.

**Comment by AndyZe** [comment]
@vatanaksoytezer can you go ahead and set the tests back as they were previously and we'll get this merged?
We're at 42 commits now. I think it would be a good idea to squash this all down to 1 commit that only updates velocity bounds enforcement.

**Comment by vatanaksoytezer** [comment]
> @vatanaksoytezer can you go ahead and set the tests back as they were previously and we'll get this merged?
> We're at 42 commits now. I think it would be a good idea to squash this all down to 1 commit that only updates velocity bounds enforcement.

@AndyZe I was waiting for today's meetup to confirm disabling tests with other people as well. Will squash and update this PR afterwards.

File: moveit/CHANGELOG.rst
Comment (tylerjw) [inline moveit/CHANGELOG.rst:9]: Why are there changelog changes in this PR?

File: moveit/package.xml
Comment (tylerjw) [inline moveit/package.xml:4]: and a version bump?

File: moveit_ros/moveit_servo/CMakeLists.txt
Comment (tylerjw) [inline moveit_ros/moveit_servo/CMakeLists.txt:254]: So this disables all the currently enabled servo tests and enables the pose tracking test?

File: moveit_ros/moveit_servo/test/config/servo_settings_low_latency.yaml
Comment (tylerjw) [inline moveit_ros/moveit_servo/test/config/servo_settings_low_latency.yaml:58]: didn't andy make another pr that added all the decial places to this?

**Review by tylerjw — COMMENTED** [review]
It seems that there are a ton of changes in this PR that are unrelated to this PR.  Please refactor those into separate PRs or just remove them from this PR if they don't belong here.

File: moveit_ros/moveit_servo/CMakeLists.txt
Comment (vatanaksoytezer) [reply moveit_ros/moveit_servo/CMakeLists.txt:254]: Yeah this seems to work that's why I left it open

File: moveit/package.xml
Comment (vatanaksoytezer) [reply moveit/package.xml:4]: I think it also included the changes in the main while rebasing I will have a look today.

File: moveit_ros/moveit_servo/test/config/servo_settings_low_latency.yaml
Comment (vatanaksoytezer) [reply moveit_ros/moveit_servo/test/config/servo_settings_low_latency.yaml:58]: > It seems that there are a ton of changes in this PR that are unrelated to this PR. Please refactor those into separate PRs or just remove them from this PR if they don't belong here.

Seems I did something wrong while squashing I will fix those today.

**Comment by vatanaksoytezer** [comment]
@AndyZe @tylerjw I would love to get a last review after squash, and merge this PR soon.

File: moveit_ros/moveit_servo/test/basic_servo_tests.cpp
Comment (AndyZe) [inline moveit_ros/moveit_servo/test/basic_servo_tests.cpp:63]: Nice change

**Review by AndyZe — APPROVED** [review]
Thanks for the squash. This looks good to me.

Please remove WIP if you think it's ready

**Comment by vatanaksoytezer** [comment]
> Thanks for the squash. This looks good to me.
> 
> Please remove WIP if you think it's ready

Removed it, thanks a lot for reviewing!

**Comment by AndyZe** [comment]
I think the reason these tests are flakey is a race condition with the launching of ros2_control and MoveIt. We saw this when developing the UR driver and some other work. Running a demo locally like this:

`ros2 launch moveit_servo servo_cpp_interface_demo.launch.py`

`ros2 control list_controllers`

I see that `panda_arm_controller` hasn't finished loading. It's in state `finalized` when it should be in state `active`. I'll make an issue to track this.

**Comment by vatanaksoytezer** [comment]
> I think the reason these tests are flakey is a race condition with the launching of ros2_control and MoveIt. We saw this when developing the UR driver and some other work. Running a demo locally like this:
> 
> `ros2 launch moveit_servo servo_cpp_interface_demo.launch.py`
> 
> `ros2 control list_controllers`
> 
> I see that `panda_arm_controller` hasn't finished loading. It's in state `finalized` when it should be in state `active`. I'll make an issue to track this.

Great catch! I think that may be the underlying problem with tests. An issue would be great.

**Comment by AndyZe** [comment]
@vatanaksoytezer it seems like I'm the only maintainer willing to approve this. Which makes sense, given the 51 comments on the PR.

I think your best bet would be to close this PR then open a new PR that _only_ ports ros-planning/moveit#2471 (no other changes). Then we could get this merged quickly, I'm fairly sure. Please!?

**Comment by vatanaksoytezer** [comment]
> @vatanaksoytezer it seems like I'm the only maintainer willing to approve this. Which makes sense, given the 51 comments on the PR.

Seems like it, yeah.

> I think your best bet would be to close this PR then open a new PR that _only_ ports [ros-planning/moveit#2471](https://github.com/ros-planning/moveit/pull/2471) (no other changes). Then we could get this merged quickly, I'm fairly sure. Please!?

Sure, I will open a new PR tomorrow when I get back to work.

**Comment by AndyZe** [comment]
Actually, @tylerjw asked for a unit test on https://github.com/ros-planning/moveit2/pull/451 that requires this PR, so we really ought to get it merged. I'll check if these tests are passing reliably yet and re-enable them if so.

**Comment by tylerjw** [comment]
I'm sorry I've been slow to respond on this.  I'll look more closely at this tonight.

**Comment by vatanaksoytezer** [comment]
Only remaining problem is: https://github.com/ros-planning/moveit2/runs/2547716859?check_suite_focus=true#step:7:7976 any ideas @tylerjw @AndyZe ? Every other test works fine and as expected. I would honestly prefer to just disable this and merge this PR and open up a new one to fix that issue.

**Comment by vatanaksoytezer** [comment]
I have disabled the tests for `servo_calcs` this will be ready to merge once the last CI passes @AndyZe @tylerjw

**Review by AndyZe — CHANGES_REQUESTED** [review]
Blocking until flaky tests are resolved

**Comment by AndyZe** [comment]
OK, we have tried very hard to fix CI. I'm at the point where we just disable the tests again.

**Review by AndyZe — APPROVED** [review]

**Comment by henningkayser** [comment]
I'm merging this, CI is fixed by #470.

**Pull request merged**
